### DISCUSSION TITLE
[RFC] Dynamic collections

### DISCUSSION THREAD

**Discussion by maarcingebala** [saleor/saleor#14408] [RFC]
The proposed solution for https://github.com/saleor/saleor/issues/10555

The goal is to allow users to create dynamic collections that automatically update the collection products based on defined rules.

## General assumptions
- The collections will be static or dynamic.
- The current collection solution will be called static collections. The flow will stay unchanged, the products will be maintained manually.
- The dynamic collections will require rules that defined the products that should be included.
- When the collection is dynamic there is no option to update it manually (adding/removing products will be blocked).
- Dynamic collection without any rules will be empty.
- The collection could be changed from dynamic to static, the products will be frozen and could be changed.
- The collection could be changed from static to dynamic - the collection products will be overwritten with products that match the rules in the asynchronous task
- After updating the collection rule the collection products will be updated in the asynchronous task.
- The celery task run periodically will be updating the collection products after any changes that may influence collections.
 
### MVP
- There will only be the possibility to define rules based on attribute values.
- The only allow operators will be `IS` and `IS NOT`, the `contains` will be not handled in the MVP version.
- The supported attribute types will be:
  - `DROPDOWN` (match by provided attribute value `id`)
  - `MULTISELECT`(match by provided attribute value `id`)
  -  `SWATCH` (match by provided attribute value `id`)
  - `PLAIN_TEXT` (only with an exact match of provided text)
  - `REFERENCE` (match by provided object reference `id`)
  - `BOOLEAN`
  - `DATE`
  - `NUMERIC`
- In the case of the multichoice the product needs to contain the given attribute value.
- The collections will be updated with products that contain specified attribute values or when any product variant fulfills rules.

### Collection recalculation
- The `Collection` will have a flag `should_refresh` that will be changed to `True` on any change that may influence the collection.
- The `CollectionCreate` and `CollectionUpdate` mutations will call the asynchronous collection recalculation for the updated collection instead of changing the flag.
- The following mutations will require the collection recalculations and will change the collection flag `should_refresh` to `True`:
  - `AttributeDelete`
  - `AttributeUpdate`
  - `AttributeValueDelete`
  - `ProductTypeUpdate` - when attributes are assigned or unassigned
  - `ProductCreate` - when any attribute values are assigned
  - `ProductUpdate` - when any attribute values changed
   - `ProductVariantCreate` - when any attribute values are assigned
  - `ProductVariantUpdate` - when any attribute values changed
- The collection recalculation will be performed in the periodic task that runs asynchronously.
- The periodic celery task should build the query that will look for all products that pass the rules. Only the collection with the `should_refresh` flag set to `True` will be updated. The product to collection will be assigned when:
  - the product has a specified attribute with the provided value,
  - or any of the product variants have the attribute with the specified value (If the attribute is multichoice, the product or variant should have at least value from the rule assigned),
  - the product fulfills all rules for a given collection.

**Questions** ❓ ❓ ❓ 
> 1. The question is: should we keep the flag in the collection or in the attribute?

In case we keep the flag in the attribute this is problematic during attribute deletion, so the flag on the collection is needed anyway.
It's easy to get the attribute-related collections, it should require only one/two additional queries, so IMO the flag on the collection should be enough.

> 2. How often the task for updating the collections should run?

Every hour or more often?

> 3. IMO, when the given collection is created or updated, the task should be run immediately. 

### Post MVP
- Add support for other operators like `CONTAINS`, `RANGE`, `DATE_RANGE`.
- Add support for other attribute types: `DATE_TIME`, `RICH_TEXT`
- Add support for the conjunction between rules like `OR` and `AND`
- Add support for different object types e.g. `ProductType`
- Add support for different types like `PRICE`, `METADATA, e.t.c.

## Database changes
Here is the proposed class diagram with new models and relations between them.
**Note:** only new fields are listed in the diagram.
```mermaid
classDiagram
  Collection "1" -- "0..*" CollectionRule
  CollectionRule "1" -- "0..*" RuleValue
  CollectionRule -- "0..1" Attribute
  RuleValue -- "0..1" AttributeValue
  class Collection {
    type: String
    should_refresh: Bool
    ...
  }
  class CollectionRule {
    qualifier: String
    operator: String
  }
  class RuleValue {
    text: String
    date: Date
    boolean: Bool
    reference_id: Int
    numeric_value: Int
  }
```
**DoD:**
- Extend the `Collection` model with `type: CharField` field with choices. The will be two options `STATIC`, and `DYNAMIC`. Set `STATIC` as default for existing and new instances.
- Extend the `Collection` with `should_refresh` field, and set `False` as default for new and existing collections.
- Add `CollectionRule` and `RuleValue` types.

```python
class CollectionType:
  STATIC
  DYNAMIC

class Collection:
  ...
  type = CharField(default=CollectionType.STATIC, choices=CollectionType)
  should_refresh = BooleanField(default=False)

class RuleOperator:
  IS
  IS_NOT

class Qualifier:
  ATTRIBUTE

class CollectionRule:
  collection = ForeignKey(Collection)
  qualifier = CharField(null=False, choices=Qualifier)
  operator = CharField(null=False, choices=RuleOperator)
  attribute =  ForeignKey(Attribute, null=True, related_name="+")
  
class RuleValue:
  collection_rule = ForeignKey(related_name="values")
  text = TextField(null=True)
  date = DateField(null=True)
  boolean = BooleanField(null=True)
  value = ForeignKey(AttributeValue, null=True, related_name="+")
  reference_id = IntegerField(null=True)
  numeric_value = IntegerField(null=True)
```

### Mutations
- Extend `CollectionCreateInput`  and `CollectionInput`  input with `type`  field. 
```graphql
enum CollectionTypeEnum {
  STATIC
  DYNAMIC
}

input CollectionCreateInput {
  ...
  type: CollectionTypeEnum
}
```
- Extend `CollectionCreateInput`  and `CollectionUpdate` with optional rules field.  Add validation in `CollectionCreate`  and `CollectionUpdate`  mutations. Raise a `ValidationError` when the rules are provided for a `STATIC`  collection.
```graphql
enum OperatorEnum {
  IS
  IS NOT
}

enum QualifierEnum {
  ATTRIBUTE
}

input CollectionRule {
  qualifier: QualifierEnum!
  operator: OperatorEnum!
  attributeRule: AttributeRuleInput
}

input AttributeRuleInput {
  attributeId: ID!
  valueIds: [ID!]
  textValue: String
  referenceIds: [ID!]
  boolean: Boolean
  dates: [Date!]
  numericValues: [Int!]
}

input CollectionCreateInput {
  ...
  rules: [CollectionRule!]
}

input CollectionInput {
  ...
  rules: [CollectionRule!]
}
```
- Add a validation, when the `ATTRIBUTE`  as a qualifier is provided: require providing the `attributeRule`.❓ Should `attributeRule`  be required as long as we have only one qualifier?
- Add validation for providing at least one of the value fields in the `AttributeRule` input.

**Post MVP:**
- `OperatorEnum` will be extended with other values e.g. `CONTAINS`, `LTE`, `GTE`.
- `QualifierEnum`  will be extended will other types values e. g. `ProductType`, and other not entity types values like `METADATA`, `PRICE` , e.t.c. 
- The CollectionRule will be extended with additional qualifier rules like `ProductTypeRule`, `MetadataRule`.

### Types
- Add a `CollectionRule` type
```graphql
type CollectionRule {
  qualifier: QualifierEnum!
  rule: Rule!
}

interface Rule {
   operator: OperatorEnum!
}

type AttributeRule  implements Rule {
  operator: OperatorEnum!
  attributeId: ID!
  valueIds: [ID!]
  textValue: String
  referenceIds: [ID!]
  boolean: Boolean
  dates: [Date!]
  numericValues: [Int!]!
}
```
>**Note**
For each new qualifier, the new `Rule` type that implements the `Rule` interface will be added.

>**Note**
When multiple values are provided the collection will be fulfilled with products that have one of the specified attribute values or their variants have.

❓ ❓ ❓ 
To consider, maybe the AttributeRule should be split into separate types for every input type:
```graphql
type DropdownAttributeRule implements Rule {
  operator: OperatorEnum!
  values: [ID!]!
}

type MultiselectAttributeRule implements Rule {
  operator: OperatorEnum!
  values: [ID!]!
}

type SwatchAttributeRule implements Rule {
  operator: OperatorEnum!
  values: [ID!]!
}

type PlainTextAttributeRule implements Rule {
  operator: OperatorEnum!
  text: String!
}

type ReferenceAttributeRule implements Rule {
  operator: OperatorEnum!
  referenceIds: [ID!]!
}

type BooleanAttributeRule implements Rule {
  operator: OperatorEnum!
  boolean: Boolean!
}

type DateAttributeRule implements Rule {
  operator: OperatorEnum!
  dates: [Date!]!
}

type NumericAttributeRule implements Rule {
  operator: OperatorEnum!
  numericValues: [Int!]!
}
``` 

- Extend the `Collection` type with `type` and `rules` field
```graphql
type Collection implements Node & ObjectWithMetadata {
  ...
  type: CollectionTypeEnum
  rules: [CollectionRule!]
}
```

**Comment by maarcingebala** [comment]
Moved from https://github.com/saleor/saleor/pull/10654

**Comment by VKalbfleisch** [comment]
Looks great and we already have some ideas on how to use it in our product. Where can I vote for prioritization of this feature?)

**Comment by peteror** [comment]
A very common collection could be everything that's part of a sale - today, we add these products to a collection manually. I see this is post MVP, but I'd consider adding it to the first version, as this is probably one of the most common use cases.
